### ISSUE TITLE
Capture a ethereum proof

### ISSUE THREAD

**Issue by ethanfrey** [confio/proofs-ethereum#1]
Look at ethereum/trie/proofs.go

This takes a "proofsDB" object with random get/set methods. To convert it, it would be much easier to have an array with multiple levels of proofs. Please write a function to query an ethereum trie for a key, and return the result in an array of steps in the ethereum native format. (This goes from root to leaf, fair enough).

Write some test code, and generate some test data. proofs_test.go is a good starting point. You can import and libraries you want here... main point is to get it to work, not reimplement rlp.

**Comment by husio** [comment]
[proof.go](https://github.com/ethereum/go-ethereum/blob/f5d89cdb72c1e82e9deb54754bef8dd20bf12591/trie/proof.go#L106)

Before I start, I want to make sure we are thinking the same way here.

> Please write a function to query an ethereum trie for a key, and return the result in an array of steps in the ethereum native format. (This goes from root to leaf, fair enough).

Should this function operate on an ethenerum database to read the data? 
- If not, where should it read the data from?
- If yes, do you have a small ethenerum database laying around? :smile:

**Comment by ethanfrey** [comment]
I have no ethereum db lying around. But you can use [this test code](https://github.com/ethereum/go-ethereum/blob/f5d89cdb72c1e82e9deb54754bef8dd20bf12591/trie/proof_test.go#L200-L223) to generate it. And then we have to use the GPL-v3 license on this repo (which is fine for any ethereum-related code)

This is a [convoluted way to make proofs](https://github.com/ethereum/go-ethereum/blob/f5d89cdb72c1e82e9deb54754bef8dd20bf12591/trie/proof_test.go#L37-L57) and then [test them](https://github.com/ethereum/go-ethereum/blob/f5d89cdb72c1e82e9deb54754bef8dd20bf12591/trie/proof_test.go#L59-L77). I think we only need to use the [direct proof format](https://github.com/ethereum/go-ethereum/blob/f5d89cdb72c1e82e9deb54754bef8dd20bf12591/trie/proof_test.go#L40-L45) and can get rid of the array of functions

**Comment by ethanfrey** [comment]
Given [Trie.Prove()](https://github.com/ethereum/go-ethereum/blob/master/trie/proof.go#L36),  I want something like this:


`func GenerateProof(trie *ethdb.Trie, key []byte) (value []byte, path []node, err error)`

Where node is a copy of [their internal types](https://github.com/ethereum/go-ethereum/blob/91eec1251c06727581063cd7e942ba913d806971/trie/node.go#L30-L47), where we can parse the binary data

(you can update the function as needed to make it cleaner)

**Comment by husio** [comment]
After digging in this for more than an hour, I think it is not possible to pull the proof value from the trie, without first implementing the protocol.

I think I will start by reading and implementing the [specification](https://github.com/ethereum/wiki/wiki/Patricia-Tree). It does not seem to be much.

I was planning to use ethereum code to build the tree and take out the hashes, so that I can use them for testing my implementation. I [wrote code](https://github.com/confio/proofs-ethereum/blob/issue_1/proof_test.go#L52) that although it made sense when I was writing it, produce unbelievable output:
```
--- PASS: TestEthTrie (0.00s)
    proof_test.go:17: empty trie root hash: 56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421
    proof_test.go:35: commit hash of the trie: 72e6c01ad0c9a7b517d4bc68a5b323287fe80f0e68f5415b4b95ecbc8ad83978
    proof_test.go:58: : node hash: 0000000000000000000000000000000000000000000000000000000000000000
    proof_test.go:58: : node hash: 72e6c01ad0c9a7b517d4bc68a5b323287fe80f0e68f5415b4b95ecbc8ad83978
    proof_test.go:58: 06: node hash: 0000000000000000000000000000000000000000000000000000000000000000
    proof_test.go:58: 0601: node hash: 0000000000000000000000000000000000000000000000000000000000000000
    proof_test.go:55: 060110: leaf "a" proof:  0 d916d780c22061c22062c2206380808080808080808080808080
    proof_test.go:58: 060110: node hash: 0000000000000000000000000000000000000000000000000000000000000000
    proof_test.go:58: 0602: node hash: 0000000000000000000000000000000000000000000000000000000000000000
    proof_test.go:55: 060210: leaf "b" proof:  0 d916d780c22061c22062c2206380808080808080808080808080
    proof_test.go:58: 060210: node hash: 0000000000000000000000000000000000000000000000000000000000000000
    proof_test.go:58: 0603: node hash: 0000000000000000000000000000000000000000000000000000000000000000
    proof_test.go:55: 060310: leaf "c" proof:  0 d916d780c22061c22062c2206380808080808080808080808080
    proof_test.go:58: 060310: node hash: 0000000000000000000000000000000000000000000000000000000000000000
PASS
```
It is quite hard to figure out something from the code as it contains no comments and those rare that are present are sometimes invalid (!!). 

Anyway, tomorrow I will try to make progress with the custom implementation. Once I have this, I hope to be able to compute the key under which the proof is stored in the ethereum database and compare implementations this way.

**Comment by ethanfrey** [comment]
I say the branch you pushed, which does start up a tree. Let me try to take a stab at it and see if I can make sense of any of this as well.

**Comment by ethanfrey** [comment]
Closed by #4
